### ISSUE TITLE
Support allocators in the standard types that we provide

### ISSUE THREAD

**Issue by griwes** [executors/futures#119]

**Comment by brycelelbach** [comment]
Note that this is just for `std::execution::semi_future<T>`, etc.

**Comment by brycelelbach** [comment]
@LeeHowes, please review, comment, discuss with @griwes and remove the discussion label when you have consensus.

@griwes, can you implement this?

**Comment by LeeHowes** [comment]
I think doing so makes sense. They will presumably need to allocate shared state so giving control of how that happens is a good idea.

Do we want to type erase the allocator?

**Comment by griwes** [comment]
I don't think we want to type erase the allocator. We can provide `pmr` aliases if someone deems that necessary. (Or people can just use `polymorphic_allocator`.)

I'll sit down to write the wording once I've given my talk.

**Comment by LeeHowes** [comment]
I think having an allocator template param on those types is fine.

When we construct off of an ContinuableFuture/SemiFuture concept implementation (or execution::promise off of a Promise) we should be clear that we don't assume that that concept also has an allocator. This would be for the shared state needed to allocate storage for that wrapping.

**Comment by dhollman** [comment]
Shouldn't this basically default to `execution::query(fut.get_executor(), execution::allocator)`? In other words, it's not really the `Future` allocating the shared state, but the executor, and that already has an allocator customization point.

I also think we should be clear that the *concepts* should **not** be required to support allocators (I know this isn't what you're suggesting here, but it's an important distinction to make). This also means that if the standard types we provide are to act as type-erasers, the type-erasing constructors, etc., should not be required to interoperate with allocators.

**Comment by griwes** [comment]
Concepts should not require supporting allocators, yes.

For the `std::execution::semi_future` and whatnot, when it's returned by the executor - sure, but sometimes you'll get it by other means, hence why `std::execution::semi_future` should support allocators for the type-erased case (and when returned from the default `query`, you could just propagate the allocator from the executor into the future) - because then it's the *future* itself that does the allocation of the shared state that does the erasure.

**Comment by dhollman** [comment]
> sure, but sometimes you'll get it by other means

I'm still not convinced of the necessity of "getting it by other means," especially if the goal of this proposal is still to provide something for executors to interoperate with.

That minor soapbox aside, though, can't you just defer any allocation until it's bound to an executor?  What allocation would need to be done for a `semi_future` that has never been bound to any executor?  (If it comes from `continuable_future::semi()`, allocation will have been done in the presence of an executor there also).

**Comment by dhollman** [comment]
Ah, good point. I was not thinking about the allocation of the type-erased wrapper itself.
